mitiq/pec/pec.py

```diff
@@ -37,6 +40,113 @@ class LargeSampleWarning(Warning):
+def intermediary_sampled_circuits(
+    circuit: QPROGRAM,
+    representations: Sequence[OperationRepresentation],
+    precision: float = 0.03,
+    num_samples: int | None = None,
+    random_state: int | np.random.RandomState | None = None,
+    return_signs: Literal[False] = False,
+) -> list[QPROGRAM]: ...
+@overload
+def intermediary_sampled_circuits(
+    circuit: QPROGRAM,
+    representations: Sequence[OperationRepresentation],
+    precision: float = 0.03,
+    num_samples: int | None = None,
+    random_state: int | np.random.RandomState | None = None,
+    *,
+    return_signs: Literal[True],
+) -> tuple[list[QPROGRAM], list[int], float, int]: ...
+def intermediary_sampled_circuits(
+    circuit: QPROGRAM,
+    representations: Sequence[OperationRepresentation],
+    precision: float = 0.03,
+    num_samples: int | None = None,
+    random_state: int | np.random.RandomState | None = None,
+    return_signs: bool = False,
+) -> list[QPROGRAM] | tuple[list[QPROGRAM], list[int], float, int]:
+    """Generates a list of sampled circuits based on the given
+    quasi-probability representations.
+
+    Args:
+        circuit: The quantum circuit to be sampled.
+        representations: The quasi-probability representations of the circuit
+            operations.
+        precision: The desired precision for the sampling process.
+            Default is 0.03.
+        num_samples: The number of samples to generate. If None, the number of
+            samples is deduced based on the precision. Default is None.
+        random_state: The random state or seed for reproducibility.
+        return_signs: If True, returns the signs and the norm along with the
+            sampled circuits. Default is False.
+
+    Returns:
+        A list of sampled circuits. If return_signs is True, also returns a
+        list of signs and the norm.
+
+    Raises:
+        ValueError: If the precision is not within the interval (0, 1].
+    """
+    if isinstance(random_state, int):
+        random_state = np.random.RandomState(random_state)
+
+    if not (0 < precision <= 1):
+        raise ValueError(
+            "The value of 'precision' should be within the interval (0, 1],"
+            f" but precision is {precision}."
+        )
+
+    # Get the 1-norm of the circuit quasi-probability representation
+    _, _, norm = sample_circuit(
+        circuit,
+        representations,
+        num_samples=1,
+    )
+
+    # Deduce the number of samples (if not given by the user)
+    if num_samples is None:
+        num_samples = int((norm / precision) ** 2)
+
+    if num_samples > 10**5:
+        warnings.warn(_LARGE_SAMPLE_WARN, LargeSampleWarning)
+
+    sampled_circuits, signs, _ = sample_circuit(
+        circuit,
+        representations,
+        random_state=random_state,
+        num_samples=num_samples,
+    )
+
+    if return_signs:
+        return sampled_circuits, signs, norm, num_samples
+    return sampled_circuits
+
+
+def combine_results(
```

Comment:
@natestemen Wouldn't you want to use `combine_results` in `execute_with_pec` as well?

Reply:
It would be nice, but because there is `full_output` option of `execute_with_pec` it would require making `combine_results` a lot more complicated so I opted not to use this function internally.

Reply:
Maybe you should add a note in this docstring that if they want all the PEC data, they cannot use this function. 
They will have to use `execute_with_PEC` with `full_output=True`.

Reply:
The user can easily compute all of the items in `pec_data` if they have the results from each circuit run.

https://github.com/unitaryfund/mitiq/blob/7e3070fd3bc1639a6125027ccaa791bee4ac608a/mitiq/pec/pec.py#L148-L156

Reply:
Yes, that's what I am trying to say. It is only available through `execute_with_pec` right now. 

Maybe direct a user to these lines for `pec_data` in the docstring of `combine_results` or when you update the user guide, show a user how to get this data by using the modularized functions. Up to you.

Reply:
I think the situation you're describing[^1] is slightly too niche for me to feel like is worth documenting. If they are aware of `full_output` it is also described in some detail in the **Returns** section of the [`execute_with_pec`](https://mitiq.readthedocs.io/en/stable/apidoc.html#mitiq.pec.pec.execute_with_pec) docs.

[^1]: 1. User is aware of the `full_output=True` option to `execute_with_pec`
    2. User decides to use circuit generation/combination workflow
    3. User wants to compute items in `full_output`, but doesn't know how.